app/controllers/host_aggregate_controller.rb

```diff
@@ -125,19 +150,261 @@ def button
+        :model   => ui_lookup(:table => 'host_aggregate'),
+        :name    => host_aggregate_name,
+        :details => task.message
+      }, :error)
+    end
+
+    @breadcrumbs.pop if @breadcrumbs
+    session[:edit] = nil
+    session[:flash_msgs] = @flash_array.dup if @flash_array
+
+    javascript_redirect :action => "show", :id => host_aggregate_id
+  end
+
+  def delete_host_aggregates
+    assert_privileges("host_aggregate_delete")
+
+    host_aggregates = if @lastaction == "show_list" || (@lastaction == "show" && @layout != "host_aggregate")
+                        find_checked_items
+                      else
+                        [params[:id]]
+                      end
+
+    if host_aggregates.empty?
+      add_flash(_("No %{models} were selected for deletion.") % {
+        :models => ui_lookup(:tables => "host_aggregate")
+      }, :error)
+    end
+
+    host_aggregates_to_delete = []
+    host_aggregates.each do |host_aggregate_id|
+      host_aggregate = HostAggregate.find_by_id(host_aggregate_id)
+      if host_aggregate.nil?
+        add_flash(_("%{model} no longer exists.") % {:model => ui_lookup(:table => "host_aggregate")}, :error)
+      elsif !host_aggregate.supports?(:delete_aggregate)
+        add_flash(_("Delete aggregate not supported by %{model} \"%{name}\"") % {
+          :model => ui_lookup(:table => 'host_aggregate'),
+          :name  => host_aggregate.name
+        }, :error)
+      else
+        host_aggregates_to_delete.push(host_aggregate)
+      end
+    end
+    process_host_aggregates(host_aggregates_to_delete, "destroy") unless host_aggregates_to_delete.empty?
+
+    # refresh the list if applicable
+    if @lastaction == "show_list"
+      show_list
+      @refresh_partial = "layouts/gtl"
+    elsif @lastaction == "show" && @layout == "host_aggregate"
+      @single_delete = true unless flash_errors?
+      if @flash_array.nil?
+        add_flash(_("The selected %{model} was deleted") % {:model => ui_lookup(:table => "host_aggregate")})
+      end
+    end
+  end
+
+  def cancel_action(message)
+    session[:edit] = nil
+    @breadcrumbs.pop if @breadcrumbs
+    javascript_redirect :action    => @lastaction,
+                        :id        => @host_aggregate.id,
+                        :display   => session[:host_aggregate_display],
+                        :flash_msg => message
   end
 
   private ############################
 
+  def form_params
+    options = {}
+    options[:name] = params[:name] if params[:name]
+    options[:availability_zone] = params[:availability_zone] if params[:availability_zone]
+    options[:ems_id] = params[:ems_id] if params[:ems_id]
+    options[:metadata] = params[:metadata] if params[:metadata]
+    options
+  end
+
+  # dispatches tasks to multiple host aggregates
+  def process_host_aggregates(host_aggregates, task)
+    return if host_aggregates.empty?
+
+    if task == "destroy"
```

Comment:
Sure. I think it's actually more readable with the nested block rather than a bunch of extra return statements,  but it's no big deal either way -- I'll change it.